api/plugin.go

```diff
@@ -0,0 +1,602 @@
+}
+
+// DestroyPluginsForUser performs teardown for all plugin instances on a particular user.
+func (c *PluginAPI) DestroyPluginsForUser(uid uint) {
+	confs := c.DB.GetPluginConfByUser(uid)
+	for _, conf := range confs {
+		if c.Holder.HasInstance(conf.Token) {
+			c.setPluginEnableStatus(conf.Token, false)
+			c.Holder.RemoveInstance(conf.Token)
+		}
+	}
+}
+
+// InitPluginsForUser instantiates all plugins for a user.
+func (c *PluginAPI) InitPluginsForUser(uid uint) {
+	u := c.DB.GetUserByID(uid)
+	c.Holder.PluginHolder.Each(func(p loader.P) {
+		c.InitPluginForUser(u, p)
+	})
+}
+
+// InitPluginForUser instantiates a plugin for a user.
+func (c *PluginAPI) InitPluginForUser(u *model.User, plugin loader.P) error {
+	userCtx := loader.UserContext{
+		ID:    u.ID,
+		Admin: u.Admin,
+		Name:  u.Name,
+	}
+	instance := plugin.GetPluginInstance(userCtx)
+	conf := c.DB.GetPluginConfByUserAndPath(u.ID, plugin.GetPluginInfo().ModulePath)
+	if conf == nil {
+		conf = new(model.PluginConf)
+		conf.UserID = u.ID
+		conf.Enabled = false
+		conf.ModulePath = plugin.GetPluginInfo().ModulePath
+		conf.Token = auth.GeneratePluginToken()
+		switch instance := instance.(type) {
+		case papiv1.Configurer:
+			conf.Config, _ = yaml.Marshal(instance.DefaultConfig())
+		}
+		c.DB.CreatePluginConf(conf)
+	}
+	if err := c.performPluginInstanceInitialization(instance, *conf); err != nil {
+		return err
+	}
+	c.Holder.AddInstance(conf.Token, instance)
+	return nil
+}
+
+func (c *PluginAPI) initPluginForAllUsers(plugin loader.P) error {
+	users := c.DB.GetUsers()
+	for _, u := range users {
+		if err := c.InitPluginForUser(u, plugin); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+type pluginStorageHandlerV1 struct {
+	mutex sync.RWMutex
+	db    PluginDatabase
+	conf  model.PluginConf
+}
+
+func (c *pluginStorageHandlerV1) getConf() error {
+	c.mutex.RLock()
+	defer c.mutex.RUnlock()
+	conf := c.db.GetPluginConfByToken(c.conf.Token)
+	if conf == nil {
+		return errors.New("configuration not found")
+	}
+	c.conf = *conf
+	return nil
+}
+
+func (c *pluginStorageHandlerV1) updateConf() error {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	return c.db.UpdatePluginConf(&c.conf)
+}
+
+func (c *pluginStorageHandlerV1) Load() ([]byte, error) {
+	if err := c.getConf(); err != nil {
+		return nil, err
+	}
+
+	return c.conf.Storage, nil
+}
+
+func (c *pluginStorageHandlerV1) Save(d []byte) error {
+	c.conf.Storage = d
+	return c.updateConf()
+}
+
+type pluginMessengerHandlerV1 struct {
+	msgAPI     MessageAPI
+	notifier   Notifier
+	db         PluginDatabase
+	conf       model.PluginConf
+	managedApp *model.Application
+}
+
+func (c *pluginMessengerHandlerV1) initializeApp(ident string) {
+	if c.conf.ApplicationToken != "" {
+		app := c.db.GetApplicationByToken(c.conf.ApplicationToken)
+		if app != nil {
+			// managed application exist
+			c.managedApp = app
+			return
+		}
+	}
+
+	app := &model.Application{
+		Token:       auth.GenerateApplicationToken(),
```

Comment:
Add a flag `Internal`-Flag to the Application if true the application should not be deletable or editable by the user. 

The internal application should not need an token as the app should only be used internally.

Reply:
The app should be editable(eg: for renaming, uploading images, etc.), but not deletable.

Reply:
right.

Reply:
I will have the API handler modified together with the WebUI change.